### PR TITLE
Finalize dashboard restrictions

### PR DESCRIPTION
**Pull request by Twixes**
## Changes

Completes last to-do of #8462 by securing the insights endpoint against unauthorized updates of insights that belong to restricted dashboards, and preventing unauthorized deletion of restricted dashboards.

## How did you test this code?

Should be covered by backend tests, but please scrutinize!

### CODE REVIEW

File: posthog/api/dashboard.py [hunk 1]
```diff
@@ -26,6 +26,15 @@
 from posthog.utils import render_template
 
 
+class CanEditDashboard(BasePermission):
```
Comment (paolodamico) [inline posthog/api/dashboard.py:29]: As a side note, feels like our permissions engine is getting quite convoluted, and will get harder to maintain. Previously I've used [DRY permissions](https://github.com/dbkaplan/dry-rest-permissions), I believe it's no longer maintained but the model was quite useful, you would add `has_object_permission` and/or `has_permission` methods to all models, and views would pull from there. That way you have a central place to understand permissions.

File: posthog/api/dashboard.py [hunk 2]
```diff
@@ -26,6 +26,15 @@
...
+    message = "This dashboard can only be edited by its owner, team members invited to editing this dashboard, and project admins."
```
Comment (paolodamico) [inline posthog/api/dashboard.py:30]: How about?
```suggestion
    message = "You don't have editing permissions for this dashboard."
```

File: posthog/api/__init__.py [hunk 3]
```diff
@@ -119,20 +117,11 @@ def api_not_found(request):
...
 projects_router.register(
     r"session_recordings", ClickhouseSessionRecordingViewSet, "project_session_recordings", ["team_id"],
 )
-
-if settings.EE_AVAILABLE:
-    from ee.clickhouse.views.insights import ClickhouseInsightsViewSet
-
-    projects_router.register(r"insights", ClickhouseInsightsViewSet, "project_insights", ["team_id"])
-else:
-    from posthog.api.insight import InsightViewSet
```
Comment (paolodamico) [inline posthog/api/__init__.py:136]: Seems like we could get rid of all these PostHog-only classes too, right?

File: ee/api/test/test_insight.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            created_by=creator,
+            restriction_level=Dashboard.RestrictionLevel.ONLY_COLLABORATORS_CAN_EDIT,
+        )
+        insight = Insight.objects.create(team=self.team, name="XYZ", created_by=self.user, dashboard=dashboard)
+
+        response = self.client.patch(f"/api/projects/{self.team.id}/insights/{insight.id}", {"name": "ABC"})
+        response_data = response.json()
+
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+        self.assertDictContainsSubset(
```
Comment (paolodamico) [inline ee/api/test/test_insight.py:24]: Nit suggestion:

```suggestion
        self.assertEquals(response_data, self.permission_denied_response("This insight is on a dashboard that can only be edited by its owner, team members invited to editing the dashboard, and project admins."))
```

File: ee/api/test/test_insight.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            {
+                "detail": "This insight is on a dashboard that can only be edited by its owner, team members invited to editing the dashboard, and project admins."
+            },
+            response_data,
+        )
```
Comment (paolodamico) [inline ee/api/test/test_insight.py:29]: Would add a quick assertion that the name was effectively not changed

**Review by paolodamico — APPROVED** [review]
lgtm!

**Review by mariusandra — APPROVED** [review]
Looks restrictive 👍 .

The only meta concern I have is that we're effectively at tradeoff junction. If we merge this in and get people using it, it'll get somewhat harder to support insights on multiple dashboards. If I recall correctly, the consensus was that we actually don't need or want that, and prefer a simpler mental model. If so, 🚢 away!

File: posthog/api/__init__.py (hunk 3 above)
Comment (Twixes) [reply posthog/api/__init__.py:136]: You mean Postgres-only? We could! Definitely for another PR though. Also, this perhaps already falls under the umbrella of @EDsCODE's yeeting, but I don't know for sure.

**Comment by Twixes** [comment]
Yeah @mariusandra There is a bit of a junction, but as we discussed, seems like shipping this now with a simple model > spending a lot of time implementing the many-to-many relationship and figuring out a more complex security model.

**Pull request merged**